ui/library/src/utils/VisibilityDetector.ts

```diff
@@ -0,0 +1,69 @@
+export type SetEntriesCallback<TEntry> = (visible: Set<TEntry>) => void;
+export type OnVisibilityChangeCallback<TEntry> = (args: {
+  visibleElements: Element[];
+  hiddenElements: Element[];
+  lastEntries: Set<TEntry>;
+}) => Set<TEntry>;
+
+export default class VisibilityDetector<TEntry> {
+  _root: Element;
+  _observer: IntersectionObserver;
+  _lastEntries: Set<TEntry>;
+  _setEntries: SetEntriesCallback<TEntry>;
+  _onVisibilityChange: OnVisibilityChangeCallback<TEntry>;
+
+  constructor({
+    root,
+    setEntries,
+    onVisibilityChange,
+  }: {
+    root: Element;
+    setEntries: SetEntriesCallback<TEntry>;
+    onVisibilityChange: OnVisibilityChangeCallback<TEntry>;
+  }) {
+    this._root = root;
+    this._lastEntries = new Set();
+    this._setEntries = setEntries;
+    this._onVisibilityChange = onVisibilityChange;
+    this._observer = new IntersectionObserver(
+      entries => {
+        // Collect visible and hidden elements
+        const visibleElements: Element[] = [];
+        const hiddenElements: Element[] = [];
+        for (const entry of entries) {
+          if (entry.isIntersecting) {
+            visibleElements.push(entry.target);
+          } else {
+            hiddenElements.push(entry.target);
+          }
+        }
+
+        // Determine what needs saved
+        const newVisibleEntries = this._onVisibilityChange({
+          visibleElements,
+          hiddenElements,
+          lastEntries: this._lastEntries,
+        });
+        const frozenEntries = new Set(newVisibleEntries);
+        Object.freeze(frozenEntries);
+        this._lastEntries = frozenEntries;
+        this._setEntries(frozenEntries);
```

Comment:
This line confuses me a bit since I feel like it is doing the same thing as line 49. `fronzenEntries` is stored two times in two different space. Also, I feel like this step can be done in the callback function `this._onVisibilityChange` on line 42. Is there any reason that we want to create a parameter for this?

Reply:
Imagine if u have 5 pages on the screen right, if the viewport is on page 2,3, our entries will be 2,3 . If u scroll to let say 3,4 what will happen is we updated the entry to 3,4 and remove out of the equation so lastEntries is technically the current position of the elements in view port and frozenEntries is just being passed to the setter which is React.useState in this scenario

Reply:
Ok. But can't we call the setter in `_onVisibilityChange()` before the callback returns the result?